frontend_tests/zjsunit/parallel_runner.js

```diff
@@ -0,0 +1,54 @@
+"use strict";
+
+const child_process = require("child_process");
+const path = require("path");
+
+const parallel = process.argv[2];
+
+// Find the files we need to run.
+const files = process.argv.slice(3);
+if (files.length === 0) {
+    throw new Error("No tests found");
+}
+
+const index = path.join(__dirname, "index.js");
+
+function create_file_groups(files, parallel) {
+    const group_size = Math.ceil(files.length / parallel);
+    const file_groups = [];
+    for (let i = 0; i < parallel; i += 1) {
+        const file_group = files.slice(i * group_size, (i + 1) * group_size);
+        file_groups.push(file_group);
+    }
+    return file_groups;
+}
+
+const file_groups = create_file_groups(files, parallel);
+const worker_processes = [];
+for (let i = 0; i < parallel; i += 1) {
+    const worker_process = child_process.spawn("node", [index, ...file_groups[i]]);
+    worker_process.on("error", (error) => {
+        // if we fail to create a child processes,
+        // exit and send a non zero return code, so we don't check coverage
+        console.error(error);
+        process.exit(2);
+    });
+    worker_process.stdout.on("data", (data) => {
+        // use process.stdout.write to avoid adding extra new lines to logs
+        process.stdout.write(String(data));
+    });
+    worker_process.stderr.on("data", (data) => {
+        // use process.stderr.write to avoid adding extra new lines to logs
+        process.stderr.write("error!\n" + String(data));
+    });
+    worker_process.on("exit", (code) => {
+        if (code === undefined || code !== 0) {
+            // if any worker_process ends without running all its test files,
+            // or due to an error, etc
+            // exit this process (and all its children) and,
+            // send a non zero return code, so we don't check coverage
+            process.exit(1);
```

Comment:
Forcing an `exit()` will leak all the remaining child processes to continue run in the background after we return to the user’s shell. Just set `process.exitCode = 1` instead.

Reply:
This comment still applies. Sending a termination signal to child processes is not the same as waiting for them to exit, so there could still be child processes running here. Use `process.exitCode`.